### PR TITLE
feat: add Input component

### LINKED ISSUE #109: Feature: Add Input Field Component
## Issue Type

- [x] Bug Report
- [x] Feature Request
- [x] Enhancement
- [ ] Other (Please specify)

## Description

Add an `Input` field component to the `primitives/Form` directory. This component is designed to be versatile and flexible, accepting a wide range of HTML form input types. Additionally, it should implement a password visible/hide feature when the input type is set to "password."

**Details:**

**Input Component:**

- **Location**: The `Input` component will be added to the `primitives/Form` directory as `input.tsx`.

- **Props**: The component should accept the following props:

  - `type`: This prop should allow any valid HTML form input type value. It provides flexibility in using different input types such as "text," "password," "email," etc.

  - `className` (optional): Developers can apply additional CSS classes as needed to customize the appearance of the input field.

  - Extending `React.InputHTMLAttributes<HTMLInputElement>`: The component's interface should extend `React.InputHTMLAttributes<HTMLInputElement>` to include standard input attributes like `placeholder`, `name`, `disabled`, etc.

- **Password Visibility Feature**: When the `type` prop is set to "password," the component should implement a password visibility toggle feature. A small eye icon can be used for this feature. To facilitate this, developers can copy the password eye icon SVG from [this Figma design](https://www.figma.com/file/WgmdhEtnF0C90qPuvvm3OS/AbegHelp-(Copy)?type=design&node-id=326-674&mode=design&t=Drn5aKdaKf6SpFy1-4).

**Outcome:**

The `Input` component will become a versatile and highly customizable form input field, supporting various input types. The added password visibility feature enhances user experience for password input fields.

**How to Implement:**

To implement this component, developers should create a new file named `input.tsx` within the `primitives/Form` directory (create the directory if it doesn't already exist). The `forwardRef` function should be used to build the component.

**Resources:**

Kindly refer to implementations in other projects for guidance:

  - [Shadcn UI Taxonomy](https://github.com/shadcn-ui/taxonomy/blob/main/components/ui/input.tsx) - `forwardRef` function implementation in a button 
  - [Boring Stack - InputField Component](https://github.com/babblebey/boring-stack/blob/feat/add-mellow-react--inputfield-component-patch/templates/mellow-react/assets/js/components/InputField/input-field.js) - Other reference including password visible/hide feature

**Note:**

Feel free to ask questions, seek clarifications, or request assistance during the implementation process.

### PR DESCRIPTION
**Pull request by Danieljs-codes**
## Description
I have implemented the creation of the Input component and also ensured cross-browser compatibility.

Please include a summary of the change and which issue is fixed (if applicable).

## Related Issue

- Fixes #109

## Contribution Guidelines

Before submitting this pull request, please review our [Contribution Guidelines](https://github.com/abeg-help/frontend/blob/dev/CONTRIBUTING.md) to understand how to contribute to this project.

## Checklist
- [x] I have reviewed the Contribution Guidelines linked above.
- [x] I have tested my changes thoroughly and ensured that all existing tests pass.
- [x] I have provided clear and concise commit messages.
- [ ] I have updated the project's documentation as necessary.

## Additional context (if needed)

https://github.com/abeg-help/frontend/assets/108600883/640ff066-b68b-41aa-9daa-09d71ed76716

### CODE REVIEW

**Review by 289Volts — CHANGES_REQUESTED** [review]
Great job. Really smart with the way you made it flexible. However, the svg should be wrapped in a button for accessibility reasons.

File: styles/globals.css
Comment (babblebey) [inline styles/globals.css:1]: Kindly revert this change... Its unrelated to this PR, I assume you made this adjustments for your testing purposes

File: pages/index.tsx
Comment (babblebey) [inline pages/index.tsx:1]: Same here, revert the changes, they're unrelated, I assume you made this adjustments for your testing purposes

**Review by babblebey — CHANGES_REQUESTED** [review]
Great stuff @Danieljs-codes 

Kindly remove the unrelated changes you've made in this PR, I assume you made them adjustments for your testing purposes.

These changes also includes the addition of the `.idea` folder.

File: components/primitives/Form/Input.tsx [hunk 1]
```diff
@@ -0,0 +1,48 @@
+import React from "react";
+import { cn } from "@/lib/utils/cn";
+
+interface InputProps extends React.InputHTMLAttributes<HTMLInputElement> {}
+
+const Input = React.forwardRef<HTMLInputElement, InputProps>(
+  ({ id, type, className, ...rest }, ref) => {
+    const [showPassword, setShowPassword] = React.useState(false);
```
Comment (babblebey) [inline components/primitives/Form/Input.tsx]: We should also do away with `state` management within `primitives`, that's the aim for grouping these category of components differently though. 

They shouldn't manage any local state at all.... You can fulfil this functionality with the `ref` instead.

File: components/primitives/Form/Input.tsx [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+      <div className="relative">
+        <input
+          className={cn(
+            "py-2 px-3 text-sm placeholder:text-sm block w-full border border-[#D0D5DD] bg-white rounded-md outline-0 focus:border-abeg-green-20 placeholder:text-abeg-neutral-50 text-abeg-neutral-10 disabled:bg-[#F0F2F5] disabled:border-[#D0D5DD]",
+            {
+              "pr-[32px]": type === "password",
+            },
+            className,
+          )}
+          type={type === "password" && showPassword ? "text" : type}
```
Comment (babblebey) [inline components/primitives/Form/Input.tsx:20]:
```suggestion
          type={type}
```

You can keep this `type` as type

File: components/primitives/Form/Input.tsx [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+        />
+        {type === "password" && (
+          <svg
+            xmlns="http://www.w3.org/2000/svg"
+            fill="none"
+            viewBox="0 0 24 24"
+            strokeWidth={1.5}
+            stroke="currentColor"
+            className="w-5 h-5 text-abeg-neutral-50 absolute left-[calc(100%-20px)] top-1/2 -translate-x-1/2 -translate-y-1/2"
+            onClick={() => setShowPassword((curPassword) => !curPassword)}
```
Comment (babblebey) [inline components/primitives/Form/Input.tsx:27]:
```suggestion
            onClick={() => ref.current.type = ref.current.type === "password" ? "text" : "password";}
```

...and you can have the `ref` change the input `type` onclick without using the `state`.

**Review by babblebey — COMMENTED** [review]
Here are some thing you can also improve on...

Kindly take note of how we intend to make the `primitives` work

File: components/primitives/Form/Input.tsx (hunk 1 above)
Comment (Danieljs-codes) [reply components/primitives/Form/Input.tsx]: ref still uses hooks Is there a reason we don't want to use state I see no difference even if we are achieving it with ref just extra stress for little to no advantage what do you think I can change it tho @289Volts

Comment (babblebey) [reply components/primitives/Form/Input.tsx]: > ..do away with `state` management within `primitives`...

We should do away with managing local states within these category of components at every opportunity that we get, that is the pattern that we are tryna adopt, haha 😃 

Kindly make the changes, it even mean less code written you see!? 🤲

File: components/primitives/Form/Input.tsx [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+            "py-2 px-3 text-sm placeholder:text-sm block w-full border border-[#D0D5DD] bg-white rounded-md outline-0 focus:border-abeg-green-20 placeholder:text-abeg-neutral-50 text-abeg-neutral-10 disabled:bg-[#F0F2F5] disabled:border-[#D0D5DD]",
+            {
+              "pr-[32px]": type === "password",
+            },
+            className,
+          )}
+          type={type === "password" && showPassword ? "text" : type}
+          ref={ref}
+          id={id}
+          {...rest}
```
Comment (babblebey) [inline components/primitives/Form/Input.tsx]:
```suggestion
          {...props}
```

please make necessary changes here too, using the term `props` is more understandable 😉

File: components/primitives/Form/Input.tsx (hunk 1 above)
Comment (Danieljs-codes) [reply components/primitives/Form/Input.tsx]: We would we be doing away with hooks but now we are changing the real DOM which according to React was meant to be an escape hatch but I will do it if that the way we are going

**Review by ryan-zayne — APPROVED** [review]
Looks good so far! Loving the flexibility aspect to the input as well👍

File: components/primitives/Form/Input.tsx (hunk 2 above)
Comment (ryan-zayne) [reply components/primitives/Form/Input.tsx:20]: For the meantime I feel this check will okay, especially when we use the input primitive for other things apart from password

**Review by 1finedev — APPROVED** [review]
LGTM

**Review by 289Volts — APPROVED** [review]

**Review by 289Volts — APPROVED** [review]

**Review by babblebey — APPROVED** [review]
